File: 1331-path-with-maximum-gold/path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        directions = [(-1, 0), (0, 1), (1, 0),(0, -1)]
        rows = len(grid)
        columns = len(grid[0])
        result = 0

        def isValid(i, j):
            if (i < 0 or j < 0 or i == rows or j == columns):
                return False
            return True

        def dfs(i, j):
            path = 0
            if (not isValid(i, j) or grid[i][j] == 0):
                return 0

            temp = grid[i][j]
            grid[i][j] = 0
            path = max(path, temp + dfs(i - 1, j)) #up
            path = max(path, temp + dfs(i, j + 1)) #right
            path = max(path, temp + dfs(i + 1, j)) #down
            path = max(path, temp + dfs(i, j - 1)) #left
            grid[i][j] = temp
            return path

        for i in range(rows):
            for j in range(columns):
                cell = grid[i][j]

                if (not cell):
                    continue

                result = max(result, dfs(i, j))

        return result

File: 1331-path-with-maximum-gold/path-with-maximum-gold.js
var getMaximumGold = function(grid) {
    let rows = grid.length
    let columns = grid[0].length
    let max_amount = 0

    function traverse(i, j) {
        let amount = 0
        if (i < 0 || j < 0 || i == rows || j == columns || grid[i][j] == 0 ) return 0
        let temp = grid[i][j]
        grid[i][j] = 0
        amount = Math.max(amount, temp + traverse(i - 1, j))
        amount = Math.max(amount, temp + traverse(i, j + 1))
        amount = Math.max(amount, temp + traverse(i + 1, j))
        amount = Math.max(amount, temp + traverse(i, j - 1))
        grid[i][j] = temp
        return amount
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            max_amount = Math.max(max_amount, traverse(i, j))
        }
    }    

    return max_amount


};